// @microsoft/sp-client-preview/lib/local/framework/shell/SuiteNavManagerConfiguration.js
/**
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * @file Represents a set of app specific configuration for the suite nav manager.
 */
"use strict";
var SuiteNavLinkIds_1 = require('../../common/suiteNav/models/SuiteNavLinkIds');
var SuiteNavManagerConfiguration = (function () {
    /**
     * Initializes a new instance of the SuiteNavManagerConfiguration.
     */
    function SuiteNavManagerConfiguration(pageContext) {
        this._pageContext = pageContext;
        this._currentUICultureName = this._pageContext.core.currentUICultureName;
        this._webServerRelativeUrl = this._pageContext.core.webServerRelativeUrl;
        this._o365ShellRenderSettings = {
            disableDelayLoad: false,
            disableShellPlus: false,
            layout: 'Mouse',
            top: undefined
        };
        this._hamburgerMenuCallback = undefined;
        this._suiteNavPostRenderCallback = undefined;
        this._systemUserKey = this._pageContext.core.systemUserKey;
        this._siteClientTag = this._pageContext.core.siteClientTag;
        this._isSuiteNavDisabled = this._shouldDisableSuiteNav();
        if (this._pageContext.core.menuData) {
            /* tslint:disable:no-string-literal */
            this._settingsLinks = this._pageContext.core.menuData['SettingsData'];
            this._signoutUrl = this._pageContext.core.menuData['SignOutUrl'];
        }
    }
    SuiteNavManagerConfiguration.prototype.isSuiteNavDisabled = function () {
        return this._isSuiteNavDisabled;
    };
    /**
     * Function for disabling the suiteNav from loading
     */
    SuiteNavManagerConfiguration.prototype.disableSuiteNav = function () {
        this._isSuiteNavDisabled = true;
    };
    /**
     * Function for modifying the suite nav data that came back from the service.
     * @param new SuiteNav data to replace the current data
     * @returns the new SuiteNav data
     */
    SuiteNavManagerConfiguration.prototype.modifySuiteNavData = function (suiteNavData) {
        // For now, always assume we're on the SharePoint home page.
        suiteNavData.CurrentMainLinkElementID = SuiteNavLinkIds_1.default.SITES;
        suiteNavData.CurrentWorkloadSettingsSubLinks =
            this._settingsLinks && this._settingsLinks.length > 0 ? this._settingsLinks : undefined;
        suiteNavData.FeedbackLink = undefined;
        if (suiteNavData.SignOutLink) {
            suiteNavData.SignOutLink.Url = this._signoutUrl;
        }
        return suiteNavData;
    };
    ;
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "systemUserKey", {
        /**
         * Gets the systemUserKey token.
         * @returns the systemUserKey or empty string
         */
        get: function () {
            return this._systemUserKey || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "cacheToken", {
        /**
         * Gets the cache token.
         * @returns the site client tag or empty string
         */
        get: function () {
            /**
             * The site client tag will be something like '0$$16.0.3912.1218' -- the only part
             * we care about is before the $$. This number will be incremented if the user's theme changes.
             */
            var clientTag = this._siteClientTag;
            if (clientTag !== undefined) {
                return clientTag.substring(0, clientTag.indexOf('$$'));
            }
            return '';
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "o365ShellRenderSettings", {
        /**
         * Gets the o365ShellRenderSettings for rendering the SuiteNav.
         * @returns _o365ShellRenderSettings
         */
        get: function () {
            return this._o365ShellRenderSettings;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "currentUICultureName", {
        /**
         * Property representing the culture name (e.g. "en-us")
         */
        get: function () {
            return this._currentUICultureName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "webServerRelativeUrl", {
        /**
         * Represents the web server relative url where we would ping the Suite Nav service.
         */
        get: function () {
            return this._webServerRelativeUrl;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "hamburgerCallback", {
        get: function () {
            return this._hamburgerMenuCallback;
        },
        set: function (hamburgerCallback) {
            this._hamburgerMenuCallback = hamburgerCallback;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "suiteNavPostRenderCallback", {
        get: function () {
            return this._suiteNavPostRenderCallback;
        },
        set: function (suiteNavPostRenderCallback) {
            this._suiteNavPostRenderCallback = suiteNavPostRenderCallback;
        },
        enumerable: true,
        configurable: true
    });
    SuiteNavManagerConfiguration.prototype._shouldDisableSuiteNav = function () {
        return (window.location.hostname === 'localhost' ||
            this._pageContext.urlQueryParameters.getValue('disableSuiteNav') === 'true') ?
            true :
            false;
    };
    return SuiteNavManagerConfiguration;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SuiteNavManagerConfiguration;
