// @microsoft/sp-client-base/lib/telemetry/Api/LogEntry.js
/**
 * @file LogEntry.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 * Represents a Log entry
 */
"use strict";
var LogFeature_1 = require('./LogFeature');
var LogType_1 = require('./LogType');
/**
 * @internal
 */
var LogEntry = (function () {
    /**
     * Initializes a new instance of LogEntry
     * @param moduleName the name of the module
     * @param logFeature the name of the feature
     * @param logType the log type
     * @param logProperties optional property bag
     */
    function LogEntry(moduleName, logFeature, logType, logProperties) {
        if (logFeature === void 0) { logFeature = LogFeature_1.default.None; }
        if (logType === void 0) { logType = LogType_1.default.Trace; }
        if (moduleName) {
            this.moduleName = moduleName;
            this.logFeature = logFeature;
            this.logType = logType;
            this.logProperties = logProperties;
        }
        else {
            throw new Error('moduleName is invalid');
        }
    }
    /**
     * convert the log entry to string
     * @returns the serialized log entry
     */
    LogEntry.prototype.toString = function () {
        var serializedLogProps = undefined;
        if (this.logProperties) {
            try {
                serializedLogProps = JSON.stringify(this.logProperties);
            }
            catch (e) {
                console.log('Error happens in toString():' + e);
            }
        }
        return this.moduleName + " | " + this.logFeature + " | " + this.logType + " | " + serializedLogProps;
    };
    return LogEntry;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = LogEntry;
