// @microsoft/sp-tslint-rules/lib/rules/a11yRoleRule.js
/**
 * @copyright Microsoft Corporation. All rights reserved.
 *
 * @a11yRole tslint rule of accessibility.
 */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Lint = require('tslint/lib/lint');
var JsxAttribute_1 = require('../utils/JsxAttribute');
// Roles definition is from https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/src/util/attributes/role.json
var roles = require('../utils/attributes/role.json');
// The array of non-abstract valid rules.
var validRoles = Object.keys(roles).filter(function (role) { return roles[role].isAbstract === false; });
function getFailureStringUndefinedRole() {
    return "'role' attribute undefined. You would like to select a role from " +
        "https://www.w3.org/TR/wai-aria/roles#role_definitions, or simply remove this attribute";
}
exports.getFailureStringUndefinedRole = getFailureStringUndefinedRole;
function getFailureStringInvalidRole(invalidRoleName) {
    return ("Invalid role attribute '" + invalidRoleName + "', elements with ARIA roles must use a valid, ") +
        "non-abstract ARIA role. A reference to role definitions can be found at " +
        "https://www.w3.org/TR/wai-aria/roles#role_definitions.";
}
exports.getFailureStringInvalidRole = getFailureStringInvalidRole;
// SOX #234814 Add optimization for applying jsx rules only on tsx files.
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        _super.apply(this, arguments);
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new A11yRoleRuleWalker(sourceFile, this.getOptions()));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var A11yRoleRuleWalker = (function (_super) {
    __extends(A11yRoleRuleWalker, _super);
    function A11yRoleRuleWalker() {
        _super.apply(this, arguments);
    }
    A11yRoleRuleWalker.prototype.visitJsxAttribute = function (node) {
        if (JsxAttribute_1.getPropName(node).toUpperCase() !== 'ROLE') {
            return;
        }
        var roleName = JsxAttribute_1.getLiteralText(node);
        if (roleName) {
            // Splitted by space doesn't mean the multiple role definition is correct,
            // just because this rule is not checking if it is using multiple role definition.
            var normalizedValues = String(roleName).toUpperCase().split(' ');
            if (normalizedValues.some(function (value) { return validRoles.indexOf(value) === -1; })) {
                this.addFailure(this.createFailure(node.getStart(), node.getWidth(), getFailureStringInvalidRole(roleName)));
            }
        }
        else if (!node.initializer ||
            roleName === '' // Check syntax like <div role='' /> or <div role="" />
        ) {
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), getFailureStringUndefinedRole()));
        }
        _super.prototype.visitJsxAttribute.call(this, node);
    };
    return A11yRoleRuleWalker;
}(Lint.RuleWalker));
