// office-ui-fabric-react/lib/demo/components/App/AppState.js
"use strict";
var BreadcrumbPage_1 = require('../../pages/BreadcrumbPage/BreadcrumbPage');
var ButtonPage_1 = require('../../pages/ButtonPage/ButtonPage');
var CalloutPage_1 = require('../../pages/CalloutPage/CalloutPage');
var ColorPickerPage_1 = require('../../pages/ColorPickerPage/ColorPickerPage');
var DocumentCardPage_1 = require('../../pages/DocumentCardPage/DocumentCardPage');
var CheckboxPage_1 = require('../../pages/CheckboxPage/CheckboxPage');
var ChoiceGroupPage_1 = require('../../pages/ChoiceGroupPage/ChoiceGroupPage');
var CommandBarPage_1 = require('../../pages/CommandBarPage/CommandBarPage');
var ContextualMenuPage_1 = require('../../pages/ContextualMenuPage/ContextualMenuPage');
var DatePickerPage_1 = require('../../pages/DatePickerPage/DatePickerPage');
var DetailsListPage_1 = require('../../pages/DetailsListPage/DetailsListPage');
var DialogPage_1 = require('../../pages/DialogPage/DialogPage');
var DropdownPage_1 = require('../../pages/DropdownPage/DropdownPage');
var FacepilePage_1 = require('../../pages/Facepile/FacepilePage');
var FocusZonePage_1 = require('../../pages/FocusZonePage/FocusZonePage');
var FocusTrapZonePage_1 = require('../../pages/FocusTrapZonePage/FocusTrapZonePage');
var GroupedListPage_1 = require('../../pages/GroupedListPage/GroupedListPage');
var ImagePage_1 = require('../../pages/ImagePage/ImagePage');
var LabelPage_1 = require('../../pages/LabelPage/LabelPage');
var LayerPage_1 = require('../../pages/LayerPage/LayerPage');
var LinkPage_1 = require('../../pages/LinkPage/LinkPage');
var ListPage_1 = require('../../pages/ListPage/ListPage');
var MarqueeSelectionPage_1 = require('../../pages/MarqueeSelectionPage/MarqueeSelectionPage');
var MessageBarPage_1 = require('../../pages/MessageBarPage/MessageBarPage');
var NavPage_1 = require('../../pages/NavPage/NavPage');
var OrgChartPage_1 = require('../../pages/OrgChartPage/OrgChartPage');
var OverlayPage_1 = require('../../pages/OverlayPage/OverlayPage');
var PanelPage_1 = require('../../pages/PanelPage/PanelPage');
var PeoplePickerPage_1 = require('../../pages/PeoplePickerPage/PeoplePickerPage');
var PersonaCardPage_1 = require('../../pages/PersonaCardPage/PersonaCardPage');
var PersonaPage_1 = require('../../pages/PersonaPage/PersonaPage');
var PivotPage_1 = require('../../pages/PivotPage/PivotPage');
var ProgressIndicatorPage_1 = require('../../pages/ProgressIndicatorPage/ProgressIndicatorPage');
var SearchBoxPage_1 = require('../../pages/SearchBoxPage/SearchBoxPage');
var SelectionPage_1 = require('../../pages/SelectionPage/SelectionPage');
var SliderPage_1 = require('../../pages/SliderPage/SliderPage');
var SpinnerPage_1 = require('../../pages/SpinnerPage/SpinnerPage');
var TextFieldPage_1 = require('../../pages/TextFieldPage/TextFieldPage');
var TogglePage_1 = require('../../pages/TogglePage/TogglePage');
var ThemePage_1 = require('../../pages/ThemePage/ThemePage');
(function (ExampleStatus) {
    ExampleStatus[ExampleStatus["placeholder"] = 0] = "placeholder";
    ExampleStatus[ExampleStatus["started"] = 1] = "started";
    ExampleStatus[ExampleStatus["beta"] = 2] = "beta";
    ExampleStatus[ExampleStatus["release"] = 3] = "release";
})(exports.ExampleStatus || (exports.ExampleStatus = {}));
var ExampleStatus = exports.ExampleStatus;
exports.AppState = {
    appTitle: 'Fabric - React',
    examplePages: [
        {
            links: [
                {
                    component: BreadcrumbPage_1.BreadcrumbPage,
                    name: 'Breadcrumb',
                    status: ExampleStatus.beta,
                    url: '#/examples/breadcrumb'
                },
                {
                    component: ButtonPage_1.ButtonPage,
                    name: 'Button',
                    status: ExampleStatus.beta,
                    url: '#/examples/button'
                },
                {
                    component: CalloutPage_1.CalloutPage,
                    name: 'Callout',
                    status: ExampleStatus.beta,
                    url: '#/examples/callout'
                },
                {
                    component: CheckboxPage_1.CheckboxPage,
                    name: 'Checkbox',
                    status: ExampleStatus.beta,
                    url: '#/examples/checkbox'
                },
                {
                    component: ChoiceGroupPage_1.ChoiceGroupPage,
                    name: 'ChoiceGroup',
                    status: ExampleStatus.beta,
                    url: '#/examples/choicegroup'
                },
                {
                    component: CommandBarPage_1.CommandBarPage,
                    name: 'CommandBar',
                    status: ExampleStatus.beta,
                    url: '#/examples/commandbar'
                },
                {
                    component: ContextualMenuPage_1.ContextualMenuPage,
                    name: 'ContextualMenu',
                    status: ExampleStatus.beta,
                    url: '#/examples/contextmenu'
                },
                {
                    component: DatePickerPage_1.DatePickerPage,
                    name: 'DatePicker',
                    status: ExampleStatus.beta,
                    url: '#/examples/datepicker'
                },
                {
                    component: DetailsListPage_1.DetailsListPage,
                    name: 'DetailsList',
                    status: ExampleStatus.beta,
                    url: '#/examples/detailslist'
                },
                {
                    component: DialogPage_1.DialogPage,
                    name: 'Dialog',
                    status: ExampleStatus.beta,
                    url: '#/examples/dialog'
                },
                {
                    component: DocumentCardPage_1.DocumentCardPage,
                    name: 'DocumentCard',
                    status: ExampleStatus.beta,
                    url: '#/examples/documentcard'
                },
                {
                    component: DropdownPage_1.DropdownPage,
                    name: 'Dropdown',
                    status: ExampleStatus.beta,
                    url: '#/examples/dropdown'
                },
                {
                    component: FacepilePage_1.FacepilePage,
                    name: 'Facepile',
                    status: ExampleStatus.started,
                    url: '#/examples/facepile'
                },
                {
                    component: LabelPage_1.LabelPage,
                    name: 'Label',
                    status: ExampleStatus.beta,
                    url: '#/examples/label'
                },
                {
                    component: LinkPage_1.LinkPage,
                    name: 'Link',
                    status: ExampleStatus.beta,
                    url: '#/examples/link'
                },
                {
                    component: ListPage_1.ListPage,
                    name: 'List',
                    status: ExampleStatus.beta,
                    url: '#/examples/list'
                },
                {
                    component: MessageBarPage_1.MessageBarPage,
                    name: 'MessageBar',
                    status: ExampleStatus.placeholder,
                    url: '#/examples/messagebar'
                },
                {
                    component: OrgChartPage_1.OrgChartPage,
                    name: 'OrgChart',
                    status: ExampleStatus.placeholder,
                    url: '#/examples/orgchart'
                },
                {
                    component: OverlayPage_1.OverlayPage,
                    name: 'Overlay',
                    status: ExampleStatus.beta,
                    url: '#/examples/overlay'
                },
                {
                    component: PanelPage_1.PanelPage,
                    name: 'Panel',
                    status: ExampleStatus.beta,
                    url: '#/examples/panel'
                },
                {
                    component: PeoplePickerPage_1.PeoplePickerPage,
                    name: 'PeoplePicker',
                    status: ExampleStatus.started,
                    url: '#/examples/peoplepicker'
                },
                {
                    component: PersonaPage_1.PersonaPage,
                    name: 'Persona',
                    status: ExampleStatus.beta,
                    url: '#/examples/persona'
                },
                {
                    component: PersonaCardPage_1.PersonaCardPage,
                    name: 'PersonaCard',
                    status: ExampleStatus.placeholder,
                    url: '#/examples/personacard'
                },
                {
                    component: PivotPage_1.PivotPage,
                    name: 'Pivot',
                    status: ExampleStatus.started,
                    url: '#/examples/pivot'
                },
                {
                    component: ProgressIndicatorPage_1.ProgressIndicatorPage,
                    name: 'ProgressIndicator',
                    status: ExampleStatus.beta,
                    url: '#/examples/progressindicator'
                },
                {
                    component: SearchBoxPage_1.SearchBoxPage,
                    name: 'SearchBox',
                    status: ExampleStatus.started,
                    url: '#/examples/searchbox'
                },
                {
                    component: SpinnerPage_1.SpinnerPage,
                    name: 'Spinner',
                    status: ExampleStatus.beta,
                    url: '#/examples/spinner'
                },
                {
                    component: TextFieldPage_1.TextFieldPage,
                    name: 'TextField',
                    status: ExampleStatus.beta,
                    url: '#/examples/textfield'
                },
                {
                    component: TogglePage_1.TogglePage,
                    name: 'Toggle',
                    status: ExampleStatus.beta,
                    url: '#/examples/toggle'
                }
            ],
            name: 'Basic components'
        },
        {
            links: [
                {
                    component: ColorPickerPage_1.ColorPickerPage,
                    name: 'ColorPicker',
                    status: ExampleStatus.started,
                    url: '#/examples/colorpicker'
                },
                {
                    component: GroupedListPage_1.GroupedListPage,
                    name: 'GroupedList',
                    status: ExampleStatus.started,
                    url: '#examples/groupedlist'
                },
                {
                    component: ImagePage_1.ImagePage,
                    name: 'Image',
                    status: ExampleStatus.beta,
                    url: '#/examples/image'
                },
                {
                    component: LayerPage_1.LayerPage,
                    name: 'Layer',
                    status: ExampleStatus.beta,
                    url: '#/examples/layer'
                },
                {
                    component: NavPage_1.NavPage,
                    name: 'Nav',
                    status: ExampleStatus.started,
                    url: '#/examples/nav'
                },
                {
                    component: SliderPage_1.SliderPage,
                    name: 'Slider',
                    status: ExampleStatus.beta,
                    url: '#/examples/slider'
                }
            ],
            name: 'Extended components'
        },
        {
            links: [
                {
                    component: FocusTrapZonePage_1.FocusTrapZonePage,
                    name: 'FocusTrapZone',
                    status: ExampleStatus.beta,
                    url: '#examples/focustrapzone'
                },
                {
                    component: FocusZonePage_1.FocusZonePage,
                    name: 'FocusZone',
                    status: ExampleStatus.beta,
                    url: '#examples/focuszone'
                },
                {
                    component: MarqueeSelectionPage_1.MarqueeSelectionPage,
                    name: 'MarqueeSelection',
                    status: ExampleStatus.beta,
                    url: '#examples/marqueeselection'
                },
                {
                    component: SelectionPage_1.SelectionPage,
                    name: 'Selection',
                    status: ExampleStatus.beta,
                    url: '#examples/selection'
                },
                {
                    component: ThemePage_1.ThemePage,
                    name: 'Themes',
                    url: '#examples/themes'
                }
            ],
            name: 'Utilities'
        }
    ],
    headerLinks: [
        {
            name: 'Getting started',
            url: '#/'
        },
        {
            name: 'Fabric',
            url: 'http://dev.office.com/fabric'
        },
        {
            name: 'Github',
            url: 'http://www.github.com/officedev'
        }
    ]
};
