// office-ui-fabric-react/lib-amd/components/ColorPicker/ColorSlider.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
define(["require", "exports", 'react', '../../utilities/css', '../../utilities/eventGroup/EventGroup'], function (require, exports, React, css_1, EventGroup_1) {
    "use strict";
    var ColorSlider = (function (_super) {
        __extends(ColorSlider, _super);
        function ColorSlider(props) {
            _super.call(this, props);
            var initialValue = this.props.initialValue;
            this._events = new EventGroup_1.EventGroup(this);
            this._onMouseDown = this._onMouseDown.bind(this);
            this._onMouseMove = this._onMouseMove.bind(this);
            this._onMouseUp = this._onMouseUp.bind(this);
            this.state = {
                isAdjusting: false,
                origin: null,
                currentValue: initialValue
            };
        }
        ColorSlider.prototype.componentWillUnmount = function () {
            this._events.dispose();
        };
        ColorSlider.prototype.render = function () {
            var _a = this.props, className = _a.className, minValue = _a.minValue, maxValue = _a.maxValue, overlayStyle = _a.overlayStyle;
            var _b = this.state, currentValue = _b.currentValue, isAdjusting = _b.isAdjusting;
            var currentPercentage = 100 * (currentValue - minValue) / (maxValue - minValue);
            return (React.createElement("div", {ref: 'root', className: css_1.css('ms-ColorPicker-slider', className, {
                'is-adjusting': isAdjusting
            }), onMouseDown: this._onMouseDown}, React.createElement("div", {className: 'ms-ColorPicker-sliderOverlay', style: overlayStyle}), React.createElement("div", {className: 'ms-ColorPicker-thumb is-slider', style: { left: currentPercentage + '%' }})));
        };
        ColorSlider.prototype._onMouseDown = function (ev) {
            this._events.on(window, 'mousemove', this._onMouseMove, true);
            this._events.on(window, 'mouseup', this._onMouseUp, true);
            this._onMouseMove(ev);
        };
        ColorSlider.prototype._onMouseMove = function (ev) {
            var _a = this.props, onChanged = _a.onChanged, minValue = _a.minValue, maxValue = _a.maxValue;
            var rectSize = this.refs.root.getBoundingClientRect();
            var currentPercentage = (ev.clientX - rectSize.left) / rectSize.width;
            var newValue = Math.min(maxValue, Math.max(minValue, currentPercentage * maxValue));
            this.setState({
                isAdjusting: true,
                currentValue: newValue
            });
            if (onChanged) {
                onChanged(newValue);
            }
            ev.preventDefault();
            ev.stopPropagation();
        };
        ColorSlider.prototype._onMouseUp = function (ev) {
            this._events.off();
            this.setState({
                isAdjusting: false,
                origin: null
            });
        };
        ColorSlider.defaultProps = {
            minValue: 0,
            maxValue: 100,
            thumbColor: 'inherit',
            initialValue: 0
        };
        return ColorSlider;
    }(React.Component));
    exports.ColorSlider = ColorSlider;
});
