// @microsoft/sp-client-preview/lib/local/clientSideWebParts/propertyPane/propertyPaneFields/propertyPaneLink/IPropertyPaneLink.js
/**
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * @file PropertyPaneLink props.
 */
"use strict";
/**
 * The position of pop up window.
 */
(function (PopupWindowPosition) {
    /**
     * PopupWindowPosition would be located in center of screen.
     */
    PopupWindowPosition[PopupWindowPosition["center"] = 0] = "center";
    /**
     * PopupWindowPosition would be located in right top of screen.
     */
    PopupWindowPosition[PopupWindowPosition["rightTop"] = 1] = "rightTop";
    /**
     * PopupWindowPosition would be located in left top of screen.
     */
    PopupWindowPosition[PopupWindowPosition["leftTop"] = 2] = "leftTop";
    /**
     * PopupWindowPosition would be located in right bottom of screen.
     */
    PopupWindowPosition[PopupWindowPosition["rightBottom"] = 3] = "rightBottom";
    /**
     * PopupWindowPosition would be located in left bottom of screen.
     */
    PopupWindowPosition[PopupWindowPosition["leftBottom"] = 4] = "leftBottom";
})(exports.PopupWindowPosition || (exports.PopupWindowPosition = {}));
var PopupWindowPosition = exports.PopupWindowPosition;
