// @microsoft/sp-client-preview/lib/local/common/webParts/PropertyPaneDefinitions.js
/**
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * @file PropertyPane definitions.
 */
"use strict";
/**
 * WebPart PropertyPane configuration events
 */
(function (WebPartConfigurationEvent) {
    WebPartConfigurationEvent[WebPartConfigurationEvent["ConfigurationComplete"] = 1] = "ConfigurationComplete";
    WebPartConfigurationEvent[WebPartConfigurationEvent["OpenComplete"] = 2] = "OpenComplete";
    WebPartConfigurationEvent[WebPartConfigurationEvent["CloseComplete"] = 3] = "CloseComplete";
    WebPartConfigurationEvent[WebPartConfigurationEvent["ApplyClicked"] = 4] = "ApplyClicked";
    WebPartConfigurationEvent[WebPartConfigurationEvent["LostFocus"] = 5] = "LostFocus";
})(exports.WebPartConfigurationEvent || (exports.WebPartConfigurationEvent = {}));
var WebPartConfigurationEvent = exports.WebPartConfigurationEvent;
/**
 * Describes the PropertyPane state.
 */
(function (PropertyPaneState) {
    // PropertyPane's open state.
    PropertyPaneState[PropertyPaneState["Open"] = 1] = "Open";
    // PropertyPane's close state.
    PropertyPaneState[PropertyPaneState["Close"] = 2] = "Close";
    // PropertyPane's Toggle state.
    PropertyPaneState[PropertyPaneState["Toggle"] = 3] = "Toggle";
    // PropertyPane's default state.
    // if it's Open it remains open with the newly selected web part's context.
    // If it's closed it stays closed.
    PropertyPaneState[PropertyPaneState["Default"] = 4] = "Default";
})(exports.PropertyPaneState || (exports.PropertyPaneState = {}));
var PropertyPaneState = exports.PropertyPaneState;
