// @microsoft/sp-client-preview/lib/local/clientSideWebParts/propertyPane/propertyPaneFields/propertyPaneField/IPropertyPaneField.js
/**
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * @file PropertyPaneField related interfaces.
 */
"use strict";
/**
 * Enum for all the supported PropertyPane field types.
 *
 * Names should be consistent with those in office-ui-fabric-react, be careful to get letter casing correct.
 */
(function (IPropertyPaneFieldType) {
    /**
     * Custom field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Custom"] = 1] = "Custom";
    /**
     * Checkbox field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["CheckBox"] = 2] = "CheckBox";
    /**
     * TextField field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["TextField"] = 3] = "TextField";
    /**
     * Toggle field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Toggle"] = 5] = "Toggle";
    /**
     * Dropdown field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Dropdown"] = 6] = "Dropdown";
    /**
     * Label field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Label"] = 7] = "Label";
    /**
     * Slider field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Slider"] = 8] = "Slider";
    /**
     * Heading field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Heading"] = 9] = "Heading";
    /**
     * Choice Group field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["ChoiceGroup"] = 10] = "ChoiceGroup";
    /**
     * Button field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Button"] = 11] = "Button";
    /**
     * Horizontal Rule field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["HorizontalRule"] = 12] = "HorizontalRule";
    /**
     * Link field.
     */
    IPropertyPaneFieldType[IPropertyPaneFieldType["Link"] = 13] = "Link";
})(exports.IPropertyPaneFieldType || (exports.IPropertyPaneFieldType = {}));
var IPropertyPaneFieldType = exports.IPropertyPaneFieldType;
