// @microsoft/sp-client-preview/lib/local/clientSideWebParts/core/test/WebPartMocks.js
"use strict";
var sp_client_base_1 = require('@microsoft/sp-client-base');
var EventAggregator_1 = require('../../../common/events/EventAggregator');
/*
 * @internal
 */
var MockWebPartManifest = (function () {
    function MockWebPartManifest() {
        this.manifestVersion = 2;
        this.id = 'wpid';
        this.componentType = 'WebPart';
        this.version = '1.0.0';
        this.preconfiguredEntries = [{
                title: { default: 'wpName' },
                description: { default: 'wpDescription' },
                officeFabricIconFontName: 'wpIcon',
                iconImageUrl: 'wpIcon',
                groupId: 'group',
                properties: {}
            }];
        this.loaderConfig = undefined;
    }
    return MockWebPartManifest;
}());
exports.MockWebPartManifest = MockWebPartManifest;
/*
 * @internal
 */
var MockWebPartManifestInstance = (function () {
    function MockWebPartManifestInstance() {
        this.manifestVersion = 2;
        this.id = 'wpid';
        this.componentType = 'WebPart';
        this.version = '1.0.0';
        this.title = { default: 'wpName' };
        this.description = { default: 'wpDescription' };
        this.officeFabricIconFontName = 'wpIcon';
        this.iconImageUrl = 'wpIcon';
        this.groupId = 'group';
        this.properties = {};
        this.loaderConfig = undefined;
    }
    return MockWebPartManifestInstance;
}());
exports.MockWebPartManifestInstance = MockWebPartManifestInstance;
/*
 * @internal
 */
var MockWebPartManagerContext = (function () {
    function MockWebPartManagerContext() {
        this.domElement = document.createElement('div');
        this.instanceId = 'wpInstanceId';
        this.manifest = new MockWebPartManifest();
        this.webPartData = undefined;
        this.displayMode = sp_client_base_1.DisplayMode.Read;
        this.host = undefined;
    }
    return MockWebPartManagerContext;
}());
exports.MockWebPartManagerContext = MockWebPartManagerContext;
/*
 * @internal
 */
var MockWebPartContext = (function () {
    function MockWebPartContext() {
        this.domElement = document.createElement('div');
        this.manifest = new MockWebPartManifestInstance();
        this.instanceId = 'wpInstanceId';
        this.webPartTag = 'WebPart.daf0b71c-6de8-4ef7-b511-faae7c388708.0ae4973a-70e3-451e-9f1f-3f25e5455a60';
        /* tslint:disable:no-any */
        this.httpClient = {};
        this.basicHttpClient = {};
        this.pageContext = {};
        this.statusRenderer = {};
        this.eventAggregator = {};
        this.configureStart = {};
        this.serviceScope = {};
        this.environment = {};
        this.host = {};
    }
    return MockWebPartContext;
}());
exports.MockWebPartContext = MockWebPartContext;
/*
 * @internal
 */
var MockPreloadedPageData = (function () {
    function MockPreloadedPageData() {
        this.appId = '00000000-0000-0000-0000-000000000000';
        this.webServerRelativeUrl = 'http://localhost:4321';
        this.webAbsoluteUrl = 'http://localhost:4321';
    }
    return MockPreloadedPageData;
}());
exports.MockPreloadedPageData = MockPreloadedPageData;
/*
 * @internal
 */
var MockWebPartHost = (function () {
    function MockWebPartHost() {
        this.pageContext = new sp_client_base_1.PageContext(undefined);
        this.basicHttpClient = undefined;
        this.httpClient = undefined;
        /* tslint:disable:no-any */
        this.eventAggregator = EventAggregator_1.default;
        /* tslint:enable:no-any */
        this.statusRenderer = new MockClientSideWebPartStatusRenderer();
        this.serviceScope = undefined;
        this.environment = new sp_client_base_1.Environment(this.serviceScope);
        this.renderPropertyPane = function (props) { };
        this.environment._initialize(sp_client_base_1.EnvironmentType.Test);
    }
    return MockWebPartHost;
}());
exports.MockWebPartHost = MockWebPartHost;
/*
 * @internal
 */
var MockClientSideWebPartStatusRenderer = (function () {
    function MockClientSideWebPartStatusRenderer() {
        this.displayCalled = false;
        this.clearCalled = false;
        this.errorCalled = false;
        this.clearErrorCalled = false;
    }
    MockClientSideWebPartStatusRenderer.prototype.displayLoadingIndicator = function (domElement) {
        this.displayCalled = true;
    };
    MockClientSideWebPartStatusRenderer.prototype.clearLoadingIndicator = function (domElement) {
        this.clearCalled = true;
    };
    MockClientSideWebPartStatusRenderer.prototype.renderError = function (domElement, error) {
        this.errorCalled = true;
    };
    MockClientSideWebPartStatusRenderer.prototype.clearError = function (domElement) {
        this.clearErrorCalled = true;
    };
    return MockClientSideWebPartStatusRenderer;
}());
exports.MockClientSideWebPartStatusRenderer = MockClientSideWebPartStatusRenderer;
