// @microsoft/sp-module-interfaces/lib/standardLibraryModules.js
"use strict";
/**
 * This is a mapping from standard library module names to their manifest IDs.
 */
exports.standardLibraryModules = {
    '@microsoft/sp-client-preview': {
        bundleReference: {
            id: '4d5eb168-6729-49a8-aec7-0e397f486b6e'
        },
        preloadId: '4d5eb168-6729-49a8-aec7-0e397f486b6e',
        dependentIds: [
            '6aeaab4a-4d96-47fe-a0c2-9e063a2efda4',
            'a8c07ea7-3ee8-4376-8afa-e0c5098615eb',
            '02a01e42-69ab-403d-8a16-acd128661f8e'
        ],
        aliases: []
    },
    'office-ui-fabric-react': {
        bundleReference: {
            id: '02a01e42-69ab-403d-8a16-acd128661f8e'
        },
        preloadId: '02a01e42-69ab-403d-8a16-acd128661f8e',
        dependentIds: ['6aeaab4a-4d96-47fe-a0c2-9e063a2efda4'],
        aliases: []
    },
    '@microsoft/sp-lodash-subset': {
        bundleReference: {
            id: '73e1dc6c-8441-42cc-ad47-4bd3659f8a3a'
        },
        preloadId: 'a8c07ea7-3ee8-4376-8afa-e0c5098615eb',
        dependentIds: [],
        aliases: [{
                id: 'a8c07ea7-3ee8-4376-8afa-e0c5098615eb',
                entryName: 'spLodash'
            }]
    },
    '@ms/sp-application-base': {
        bundleReference: {
            id: '4df9bb86-ab0a-4aab-ab5f-48bf167048fb'
        },
        preloadId: '4df9bb86-ab0a-4aab-ab5f-48bf167048fb',
        dependentIds: [],
        aliases: []
    },
    '@ms/sp-canvas': {
        bundleReference: {
            id: '85093aa7-8c12-4683-91aa-47cd5e2654db'
        },
        preloadId: '85093aa7-8c12-4683-91aa-47cd5e2654db',
        dependentIds: [],
        aliases: []
    },
    '@microsoft/sp-client-base': {
        bundleReference: {
            id: 'af59c2b3-2da7-41fd-8b72-3939817960af'
        },
        preloadId: 'a8c07ea7-3ee8-4376-8afa-e0c5098615eb',
        dependentIds: [],
        aliases: [{
                id: 'a8c07ea7-3ee8-4376-8afa-e0c5098615eb',
                entryName: 'spClientBase'
            }]
    },
    '@ms/odsp-utilities-bundle': {
        bundleReference: {
            id: 'cc2cc925-b5be-41bb-880a-f0f8030c6aff'
        },
        preloadId: 'a8c07ea7-3ee8-4376-8afa-e0c5098615eb',
        dependentIds: [],
        aliases: [{
                id: 'a8c07ea7-3ee8-4376-8afa-e0c5098615eb',
                entryName: 'odspUtilitiesBundle'
            }]
    },
    '@ms/sp-webpart-base': {
        bundleReference: {
            id: '974a7777-0990-4136-8fa6-95d80114c2e0'
        },
        preloadId: '974a7777-0990-4136-8fa6-95d80114c2e0',
        dependentIds: [],
        aliases: []
    },
    'react': {
        bundleReference: {
            id: '6aeaab4a-4d96-47fe-a0c2-9e063a2efda4',
            entryName: 'react'
        },
        preloadId: '6aeaab4a-4d96-47fe-a0c2-9e063a2efda4',
        dependentIds: [],
        aliases: []
    },
    'react-dom': {
        bundleReference: {
            id: '6aeaab4a-4d96-47fe-a0c2-9e063a2efda4',
            entryName: 'reactDom'
        },
        preloadId: '6aeaab4a-4d96-47fe-a0c2-9e063a2efda4',
        dependentIds: [],
        aliases: []
    },
    'flux': {
        bundleReference: {
            id: '6aeaab4a-4d96-47fe-a0c2-9e063a2efda4',
            entryName: 'flux'
        },
        preloadId: '6aeaab4a-4d96-47fe-a0c2-9e063a2efda4',
        dependentIds: [],
        aliases: []
    },
    '@ms/i18n-utilities': {
        bundleReference: {
            id: 'c83d5509-ccd5-4c67-919f-2440f237927a'
        },
        preloadId: 'c83d5509-ccd5-4c67-919f-2440f237927a',
        dependentIds: [],
        aliases: []
    }
};
