// @microsoft/sp-module-loader/lib/DeveloperTools/Components/DeveloperToolsConsole.js
/**
 * @file DeveloperToolsConsole.tsx
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var React = require('react');
var DeveloperToolsConsoleActionCreator_1 = require('./../ActionCreators/DeveloperToolsConsoleActionCreator');
var DeveloperToolsConsoleStore_1 = require('./../Stores/DeveloperToolsConsoleStore');
var ManifestManager_1 = require('./DeveloperModules/ManifestManager/ManifestManager');
var PerformanceDisplay_1 = require('./DeveloperModules/PerformanceDisplay/PerformanceDisplay');
var Tabs_1 = require('./Tabs/Tabs');
var DeveloperToolsConsole_module_scss_1 = require('./DeveloperToolsConsole.module.scss');
/**
 * @todo (VSO bug 236724): localize these strings
 * Const variables for the tab keys. Used to help keep things consistent and clean when the number of tabs grows
 */
var manifests = 'manifests';
var performance = 'performance';
var tabs = (_a = {},
    _a[manifests] = 'Manifests',
    _a[performance] = 'Performance',
    _a
);
var DeveloperToolsConsole = (function (_super) {
    __extends(DeveloperToolsConsole, _super);
    function DeveloperToolsConsole() {
        var _this = this;
        _super.call(this);
        this._onDrag = function (e) {
            /* tslint:disable:no-any */
            var mouseEvent = e;
            /* tslint:enable:no-any */
            if (mouseEvent.pageY !== 0) {
                var height = (document.documentElement.clientHeight - mouseEvent.pageY) /
                    document.documentElement.clientHeight;
                DeveloperToolsConsoleActionCreator_1.default.resizeConsole(height);
            }
        };
        this._setState = function () {
            _this.setState({
                topPosition: DeveloperToolsConsoleStore_1.default.height,
                visible: DeveloperToolsConsoleStore_1.default.visible,
                selectedTabId: DeveloperToolsConsoleStore_1.default.selectedTabId
            });
        };
        this.state = {
            topPosition: DeveloperToolsConsoleStore_1.default.height,
            visible: DeveloperToolsConsoleStore_1.default.visible,
            selectedTabId: DeveloperToolsConsoleStore_1.default.selectedTabId
        };
    }
    DeveloperToolsConsole.prototype.componentDidMount = function () {
        DeveloperToolsConsoleStore_1.default.addListener(this._setState);
    };
    DeveloperToolsConsole.prototype.componentWillUnmount = function () {
        DeveloperToolsConsoleStore_1.default.removeListener(this._setState);
    };
    DeveloperToolsConsole.prototype.render = function () {
        if (this.state.visible) {
            var developerModule = void 0;
            switch (this.state.selectedTabId) {
                case manifests:
                    developerModule = React.createElement(ManifestManager_1.default, null);
                    break;
                case performance:
                    developerModule = React.createElement(PerformanceDisplay_1.default, null);
                    break;
                default:
                    developerModule = React.createElement("div", {className: DeveloperToolsConsole_module_scss_1.default.notImplementedModule}, "NOT IMPLEMENTED");
                    break;
            }
            return React.createElement("div", {className: DeveloperToolsConsole_module_scss_1.default.container, style: { height: this.state.topPosition * 100 + "%" }}, React.createElement("div", {className: DeveloperToolsConsole_module_scss_1.default.resizeBar, draggable: true, onDrag: this._onDrag}), React.createElement("div", null, React.createElement("button", {onClick: function () { return DeveloperToolsConsoleActionCreator_1.default.showHideConsole(false); }, className: [DeveloperToolsConsole_module_scss_1.default.closeButton, 'ms-Icon--x', 'ms-Icon'].join(' ')}), React.createElement(Tabs_1.default, {tabs: tabs, selectedId: this.state.selectedTabId, selectionHandler: function (id) { return DeveloperToolsConsoleActionCreator_1.default.selectTab(id); }}), React.createElement("div", {className: DeveloperToolsConsole_module_scss_1.default.developerModuleArea}, developerModule)));
        }
        else {
            /* tslint:disable:no-null-keyword */
            return null;
        }
    };
    return DeveloperToolsConsole;
}(React.Component));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = DeveloperToolsConsole;
var _a;
