// @microsoft/sp-build-core-tasks/lib/index.js
"use strict";
// Imports
var gulp_core_build_1 = require('gulp-core-build');
var ConfigureWebpackTask_1 = require('./configureWebpack/ConfigureWebpackTask');
var CopyAssetsTask_1 = require('./copyAssets/CopyAssetsTask');
var WriteManifestsTask_1 = require('./writeManifests/WriteManifestsTask');
var CasperJSTask_1 = require('./casprJs/CasperJSTask');
var CopyStaticAssetsTask_1 = require('./copyStaticAssets/CopyStaticAssetsTask');
var CollectLocalizedResourcesTask_1 = require('./collectLocalizedResources/CollectLocalizedResourcesTask');
var PackageSolutionTask_1 = require('./packageSolution/PackageSolutionTask');
var DeployAzureStorageTask_1 = require('./deployAzureStorage/DeployAzureStorageTask');
// Exports
exports.configureWebpack = new ConfigureWebpackTask_1.ConfigureWebpackTask();
exports.copyAssets = new CopyAssetsTask_1.CopyAssetsTask();
exports.writeManifests = new WriteManifestsTask_1.WriteManifestsTask();
exports.casperJSRunner = new CasperJSTask_1.CasperJSTask();
exports.copyStaticAssets = new CopyStaticAssetsTask_1.CopyStaticAssetsTask();
exports.collectLocalizedResources = new CollectLocalizedResourcesTask_1.CollectLocalizedResourcesTask();
exports.packageSolution = new PackageSolutionTask_1.PackageSolutionTask();
exports.deployAzureStorage = new DeployAzureStorageTask_1.DeployAzureStorageTask();
// @todo (SPPPlat VSO bug #189028) This is a workaround for a TypeScript compiler bug
// that we reported.  After the compiler is fixed, we will remove the typecast.
/* tslint:disable:no-any */
exports.prepareDeploy = gulp_core_build_1.serial(exports.collectLocalizedResources, exports.copyAssets);
/* tslint:disable:no-any */
