// @microsoft/sp-build-core-tasks/lib/packageSolution/packageSolution/xmlGenerators/createWebPartXml.js
/**
 * @file createwebPartFile.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * Creates an .WebPart definition for a Client-Side WebPart
 */
"use strict";
/// <reference path='./xml.d.ts' />
var path = require('path');
var XmlGenerator = require('xml');
var serializationOptions_1 = require('./serializationOptions');
// @todo: (SPPPLAT VSO#224409) Need to cleanup this implementation to generate as many .webpart
// files as preconfiguredEntries and remove any's.
function createWebPartXml(webpart, featureId) {
    /* tslint:disable:no-any */
    var webPartManifest = JSON.parse(webpart.manifest);
    /* tslint:any:no-any */
    var createPropertyDefinition = function (name, type, value) {
        return {
            property: [
                {
                    _attr: {
                        name: name,
                        type: type
                    }
                },
                value
            ]
        };
    };
    return {
        filename: path.posix.join(featureId, "WebPart_" + webpart.id + ".webpart"),
        xml: XmlGenerator({
            webParts: [
                {
                    webPart: [
                        {
                            _attr: {
                                xmlns: 'http://schemas.microsoft.com/WebPart/v3'
                            }
                        },
                        {
                            metaData: [
                                {
                                    type: [
                                        {
                                            _attr: {
                                                /* tslint:disable:max-line-length */
                                                name: 'Microsoft.SharePoint.WebPartPages.ClientSideWebPart, Microsoft.SharePoint, Version=16.0.0.0, Culture=neutral, PublicKeyToken=71e9bce111e9429c'
                                            }
                                        }
                                    ]
                                },
                                {
                                    importErrorMessage: 'Cannot import this Web Part'
                                }
                            ]
                        },
                        {
                            data: [
                                {
                                    properties: [
                                        createPropertyDefinition('Title', 'string', webPartManifest.preconfiguredEntries[0].title.default),
                                        createPropertyDefinition('Description', 'string', webPartManifest.preconfiguredEntries[0].description.default),
                                        createPropertyDefinition('ClientSideWebPartId', '', webpart.id),
                                        createPropertyDefinition('ClientSideWebPartData', 'string', '')
                                    ]
                                }
                            ]
                        }
                    ]
                }
            ]
        }, serializationOptions_1.default)
    };
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = createWebPartXml;
