// @microsoft/sp-client-preview/lib/local/framework/pageContext/PageContextItem.js
"use strict";
/**
  * This enum is used with the PageContextItem.state property to indicate
  * whether the PageContext objects are ready for usage.,
  */
(function (PageContextItemState) {
    /**
     * The object is not initialized, and there is no pending request to load it.
     */
    PageContextItemState[PageContextItemState["Uninitialized"] = 0] = "Uninitialized";
    /**
     * A promise is currently executing.  When it completes, the object will be loaded.
     */
    PageContextItemState[PageContextItemState["Requested"] = 1] = "Requested";
    /**
     * The object has been loaded and its state is valid.
     */
    PageContextItemState[PageContextItemState["Ready"] = 2] = "Ready";
})(exports.PageContextItemState || (exports.PageContextItemState = {}));
var PageContextItemState = exports.PageContextItemState;
/**
 * This is an abstract base class for the SPSite, SPWeb, SPList, SPListItem,
 * and ApplicationContext classes.  It defines the "state" of these objects.
 */
var PageContextItem = (function () {
    function PageContextItem(pageContext) {
        this._pageContext = pageContext;
        this._state = PageContextItemState.Uninitialized;
    }
    /**
     * Returns the PageContext that owns this object.  You can use this to find
     * other related objects, e.g. the parent SPWeb for an SPList.
     */
    PageContextItem.prototype.getPageContext = function () {
        return this._pageContext;
    };
    Object.defineProperty(PageContextItem.prototype, "state", {
        /**
         * Indicates whether the context item is uninitialized, waiting to be loaded,
         * or ready to use.
         */
        get: function () {
            return this._state;
        },
        set: function (value) {
            this._state = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Reports an error if the object has already been loaded.
     */
    PageContextItem.prototype.validateLoad = function () {
        if (this._state === PageContextItemState.Ready) {
            throw new Error('The object has already been loaded');
        }
    };
    return PageContextItem;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PageContextItem;
