// @microsoft/sp-client-preview/lib/local/framework/pageContext/SPList.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var PageContextItem_1 = require('./PageContextItem');
var GuidHelpers_1 = require('../basicTypes/GuidHelpers');
/**
  * This class is primarily used with the PageContext class.  It provides contextual
  * information for the SharePoint site collection that is hosting the page.
  *
  * @alpha
  */
var SPList = (function (_super) {
    __extends(SPList, _super);
    function SPList(pageContext) {
        _super.call(this, pageContext);
        this._id = undefined;
        this._title = undefined;
        this._description = undefined;
    }
    /**
     * Assigns all properties of the object, then assigns the "state" property
     * to PageContextItemState.Ready.
     */
    SPList.prototype.load = function (args) {
        this.validateLoad();
        GuidHelpers_1.default.requireValid(args.id);
        this._id = args.id;
        this._title = args.title;
        this._description = args.description;
        this.state = PageContextItem_1.PageContextItemState.Ready;
    };
    Object.defineProperty(SPList.prototype, "id", {
        /**
         * The GUID that identifies the SPList on the server.
         */
        get: function () {
            return this._id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPList.prototype, "title", {
        /**
         * The title of the list.
         */
        get: function () {
            return this._title;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPList.prototype, "description", {
        /**
         * The description of the list.
         */
        get: function () {
            return this._description;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPList.prototype, "parentWeb", {
        /**
         * The parent SPWeb that this list belongs to.
         */
        get: function () {
            return this.getPageContext().web;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns an ODATA object containing additional information for this list, if available.
     * Otherwise, this returns undefined.  If your application fetches this data itself, it
     * can call updateOdataObject() to update the context.
     */
    SPList.prototype.getOdataObject = function () {
        return this._odataObject;
    };
    /**
     * Updates the ODATA object that will be returned when getOdataObject() is called.
     * This allows the application to share its data with other objects (e.g. web parts)
     * to avoid unnecessary REST calls.
     */
    SPList.prototype.updateOdataObject = function (value) {
        this._odataObject = value;
    };
    return SPList;
}(PageContextItem_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPList;
