// @microsoft/sp-client-preview/lib/local/framework/pageContext/SPListItem.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var PageContextItem_1 = require('./PageContextItem');
/**
  * This class is primarily used with the PageContext class.  It provides contextual
  * information for the SharePoint list item associated with the current page.
  *
  * @alpha
  */
var SPListItem = (function (_super) {
    __extends(SPListItem, _super);
    function SPListItem(pageContext) {
        _super.call(this, pageContext);
        this._id = undefined;
        this._list = undefined;
        this._odataObject = undefined;
    }
    /**
     * Assigns all properties of the object, then assigns the "state" property
     * to PageContextItemState.Ready.
     */
    SPListItem.prototype.load = function (args) {
        this.validateLoad();
        this._id = args.id;
        this.state = PageContextItem_1.PageContextItemState.Ready;
    };
    Object.defineProperty(SPListItem.prototype, "id", {
        /**
         * Returns the integer that identifies this list item on the server.  These integers
         * are assigned in increasing order.
         */
        get: function () {
            return this._id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPListItem.prototype, "parentList", {
        /**
         * Returns the SPList that this list item belongs to.
         */
        get: function () {
            return this.getPageContext().list;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns an ODATA object containing additional information for this list item, if available.
     * Otherwise, this returns undefined.  If your application fetches this data itself, it
     * can call updateOdataObject() to update the context.
     */
    SPListItem.prototype.getOdataObject = function () {
        return this._odataObject;
    };
    /**
     * Updates the ODATA object that will be returned when getOdataObject() is called.
     * This allows the application to share its data with other objects (e.g. web parts)
     * to avoid unnecessary REST calls.
     */
    SPListItem.prototype.updateOdataObject = function (value) {
        this._odataObject = value;
    };
    return SPListItem;
}(PageContextItem_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPListItem;
