// @microsoft/sp-client-preview/lib/local/framework/pageContext/SPNavigationNode.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var PageContextItem_1 = require('./PageContextItem');
/**
  * This class is primarily used with the PageContext class.  It provides quick launch
  * navigation information for the hosting application.
  *
  * @alpha
  */
var SPNavigationNode = (function (_super) {
    __extends(SPNavigationNode, _super);
    function SPNavigationNode(pageContext) {
        _super.call(this, pageContext);
    }
    /**
     * Assigns all properties of the object, then assigns the "state" property
     * to PageContextItemState.Ready.
     */
    SPNavigationNode.prototype.load = function (args) {
        this.validateLoad();
        this._id = args.id;
        this.state = PageContextItem_1.PageContextItemState.Ready;
    };
    Object.defineProperty(SPNavigationNode.prototype, "id", {
        /**
         * The numeric ID that identifies the SPNavigationNode on the server.
         */
        get: function () {
            return this._id;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns an ODATA object containing additional navigation information if available.
     * Otherwise, this returns undefined.  If your application fetches thiscls
     * data itself, it can call updateOdataObject() to update the context.
     */
    SPNavigationNode.prototype.getOdataObject = function () {
        return this._odataObject;
    };
    /**
     * Updates the ODATA object that will be returned when getOdataObject() is called.
     * This allows the application to share its data with other objects (e.g. web parts)
     * to avoid unnecessary REST calls.
     */
    SPNavigationNode.prototype.updateOdataObject = function (value) {
        this._odataObject = value;
    };
    return SPNavigationNode;
}(PageContextItem_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPNavigationNode;
