// @microsoft/sp-client-preview/lib/local/framework/pageContext/SPUser.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var PageContextItem_1 = require('./PageContextItem');
/**
  * This class is primarily used with the PageContext class.  It provides contextual
  * information for the current user visiting the page.
  *
  * @alpha
  */
var SPUser = (function (_super) {
    __extends(SPUser, _super);
    function SPUser(pageContext) {
        _super.call(this, pageContext);
        this._id = undefined;
    }
    /**
     * Assigns all properties of the object, then assigns the "state" property
     * to PageContextItemState.Ready.
     */
    SPUser.prototype.load = function (args) {
        this.validateLoad();
        this._id = args.id;
        this.state = PageContextItem_1.PageContextItemState.Ready;
    };
    Object.defineProperty(SPUser.prototype, "id", {
        /**
         * The ID that identifies the SPUser on the server.
         */
        get: function () {
            return this._id;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns an ODATA object containing additional information for this user, if available.
     * Otherwise, this returns undefined.  If your application fetches this data itself, it
     * can call updateOdataObject() to update the context.
     */
    SPUser.prototype.getOdataObject = function () {
        return this._odataObject;
    };
    /**
     * Updates the ODATA object that will be returned when getOdataObject() is called.
     * This allows the application to share its data with other objects (e.g. web parts)
     * to avoid unnecessary REST calls.
     */
    SPUser.prototype.updateOdataObject = function (value) {
        this._odataObject = value;
    };
    return SPUser;
}(PageContextItem_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPUser;
