// @microsoft/sp-client-preview/lib/local/framework/pageContext/SPWeb.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var PageContextItem_1 = require('./PageContextItem');
var url_1 = require('../basicTypes/url');
var GuidHelpers_1 = require('../basicTypes/GuidHelpers');
/**
  * This class is primarily used with the PageContext class.  It provides contextual
  * information for the SharePoint site ("web") that hosts the page.
  *
  * @alpha
  */
var SPWeb = (function (_super) {
    __extends(SPWeb, _super);
    function SPWeb(pageContext) {
        _super.call(this, pageContext);
        this._id = undefined;
        this._serverRelativeUrl = undefined;
        this._title = undefined;
    }
    /**
     * Assigns all properties of the object, then assigns the "state" property
     * to PageContextItemState.Ready.
     */
    SPWeb.prototype.load = function (args) {
        this.validateLoad();
        GuidHelpers_1.default.requireValid(args.id);
        this._id = args.id;
        url_1.requireServerRelativeUrl(args.serverRelativeUrl);
        this._serverRelativeUrl = args.serverRelativeUrl;
        this._title = args.title;
        this.state = PageContextItem_1.PageContextItemState.Ready;
    };
    Object.defineProperty(SPWeb.prototype, "id", {
        /**
         * The GUID that identifies the SPWeb on the server.
         */
        get: function () {
            return this._id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPWeb.prototype, "serverRelativeUrl", {
        /**
         * Returns a server-relative URL for this SPWeb.
         * Example: "/sites/PubSite/SubWeb"
         */
        get: function () {
            return this._serverRelativeUrl;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPWeb.prototype, "title", {
        /**
         * Returns the title of the SharePoint site.
         */
        get: function () {
            return this._title;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPWeb.prototype, "parentSite", {
        /**
         * Returns the parent site collection that this site ("web") belongs to.
         */
        get: function () {
            return this.getPageContext().site;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns an ODATA object containing additional information for this SharePoint site ("web"),
     * if available.  Otherwise, this returns undefined.  If your application fetches this
     * data itself, it can call updateOdataObject() to update the context.
     */
    SPWeb.prototype.getOdataObject = function () {
        return this._odataObject;
    };
    /**
     * Updates the ODATA object that will be returned when getOdataObject() is called.
     * This allows the application to share its data with other objects (e.g. web parts)
     * to avoid unnecessary REST calls.
     */
    SPWeb.prototype.updateOdataObject = function (value) {
        this._odataObject = value;
    };
    return SPWeb;
}(PageContextItem_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPWeb;
