// @microsoft/sp-client-base/lib/oauthTokenProvider/OAuthTokenProvider.js
"use strict";
var sp_client_shared_1 = require('@ms/sp-client-shared');
var OAuthUtilityContext_1 = require('./OAuthUtilityContext');
var ServiceKey_1 = require('../serviceScope/ServiceKey');
var Validate_1 = require('../common/Validate');
/**
 * This class allows one to obtain OAuth tokens
 *
 * OAuth tokens are used to authenticate the user from the sharepoint page
 * to other services such as PowerBI, Sway, Exchange, Yammer, etc.
 *
 * @alpha
 */
var OAuthTokenProvider = (function () {
    function OAuthTokenProvider(serviceScope) {
        this._tokenUtility = new sp_client_shared_1.OAuthUtility(new OAuthUtilityContext_1.default(serviceScope));
    }
    /**
     * Fetches the oauth token for a resource if the user that's currently logged in has
     * access to that resource.
     *
     * The OAuth token SHOULD NOT be cached by the caller since it is already cached by the method
     * itself.
     *
     * If the promise returned by this method rejects, there is no point in retrying the same call.
     * If the response from the server signals that the request should be retried this function will do this by itself.
     *
     * An example of a resource would be https://sdfpilot.outlook.com
     *
     * @param resource - the resource for which the token should be obtained
     * @returns A promise that will be fullfiled with the token or that will reject
     *          with an instance of TokenUtilityError
     */
    // @todo VSO:237921 - use the correct typings
    OAuthTokenProvider.prototype.getOAuthToken = function (resource) {
        Validate_1.default.isNotNullOrUndefined(resource, 'resource');
        Validate_1.default.isNonemptyString(resource, 'resource');
        return this._tokenUtility.getOAuthToken(resource);
    };
    /**
     * Clears the token from a given resource from the internal cache. If there is no token
     * cached for the resource the method does nothing.
     *
     * @param resource - resource for which the token should be cleared from the cache
     **/
    OAuthTokenProvider.prototype.clearCachedToken = function (resource) {
        Validate_1.default.isNotNullOrUndefined(resource, 'resource');
        Validate_1.default.isNonemptyString(resource, 'resource');
        this._tokenUtility.clearCachedToken(resource);
    };
    return OAuthTokenProvider;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = OAuthTokenProvider;
exports.oAuthTokenProviderServiceKey = ServiceKey_1.default.create('sp-client-shared:OAuthTokenProvider', OAuthTokenProvider);
