// @microsoft/sp-client-base/lib/oauthTokenProvider/OAuthUtilityContext.js
"use strict";
var QosMonitor_1 = require('../telemetry/Api/QosMonitor');
var LogEvent_1 = require('../telemetry/Api/LogEvent');
var LogLevel_1 = require('../telemetry/Api/LogLevel');
var HttpClient_1 = require('../httpClient/HttpClient');
var TokenUtilityLogger = (function () {
    function TokenUtilityLogger() {
    }
    TokenUtilityLogger.prototype.logInfo = function (message) {
        LogEvent_1.default.log(TokenUtilityLogger.logSource, message, LogLevel_1.LogLevel.Info);
    };
    TokenUtilityLogger.prototype.logError = function (message) {
        LogEvent_1.default.log(TokenUtilityLogger.logSource, message, LogLevel_1.LogLevel.Error);
    };
    TokenUtilityLogger.prototype.logWarning = function (message) {
        LogEvent_1.default.log(TokenUtilityLogger.logSource, message, LogLevel_1.LogLevel.Warning);
    };
    TokenUtilityLogger.logSource = 'TokenUtility';
    return TokenUtilityLogger;
}());
/**
 * The DelegationTokenProvider's implementation was moved into the sp-client-shared
 * package so that it can be shared with other teams that are not using sp-client-base
 * platform.  To support this, the sp-client-shared package defines an interface
 * ITokenUtilityContext which abstracts the basic services that are needed by that
 * code.  TokenUtilityContext is our adapter that wires up this interface to the
 * corresponding services in sp-client-base; it can be considered part of the
 * internal implementation of DelegationTokenProvider; it is not  intended to be
 * used by other parts of our framework.
 */
var OAuthUtilityContext = (function () {
    function OAuthUtilityContext(serviceScope) {
        var _this = this;
        serviceScope.whenFinished(function () {
            _this._httpClient = serviceScope.consume(HttpClient_1.httpClientServiceKey);
            _this._logger = new TokenUtilityLogger();
        });
    }
    Object.defineProperty(OAuthUtilityContext.prototype, "logger", {
        get: function () {
            return this._logger;
        },
        enumerable: true,
        configurable: true
    });
    OAuthUtilityContext.prototype.fetchWithDigest = function (request) {
        var _this = this;
        return request.blob().then(function (requestBody) {
            return _this._httpClient.fetch(request.url, {
                method: request.method,
                headers: request.headers,
                body: requestBody
            });
        });
    };
    OAuthUtilityContext.prototype.qosMonitor = function () {
        return new QosMonitor_1.default(OAuthUtilityContext.qosName);
    };
    OAuthUtilityContext.qosName = 'OAuthTokenProvider';
    return OAuthUtilityContext;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = OAuthUtilityContext;
