// @microsoft/sp-module-loader/lib/DeveloperTools/Components/DeveloperModules/ManifestManager/ManifestManager.js
/**
 * @file ManifestManager.tsx
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var React = require('react');
var ManifestManagerStore_1 = require('./../../../Stores/ManifestManagerStore');
var ManifestManagerActionCreator_1 = require('./../../../ActionCreators/ManifestManagerActionCreator');
var ManifestList_1 = require('./ManifestList/ManifestList');
var ManifestDetails_1 = require('./ManifestDetails/ManifestDetails');
var ManifestManager_module_scss_1 = require('./ManifestManager.module.scss');
var ManifestManager = (function (_super) {
    __extends(ManifestManager, _super);
    function ManifestManager() {
        var _this = this;
        _super.call(this);
        this._setState = function () {
            _this.setState({
                manifests: ManifestManagerStore_1.default.registeredManifests,
                selectedManifestId: ManifestManagerStore_1.default.selectedManifestId
            });
        };
        this.state = {
            manifests: ManifestManagerStore_1.default.registeredManifests,
            selectedManifestId: ManifestManagerStore_1.default.selectedManifestId
        };
    }
    ManifestManager.prototype.componentDidMount = function () {
        ManifestManagerStore_1.default.addListener(this._setState);
    };
    ManifestManager.prototype.componentWillUnmount = function () {
        ManifestManagerStore_1.default.removeListener(this._setState);
    };
    ManifestManager.prototype.render = function () {
        var _this = this;
        var selectedManifest = this.state.manifests[this.state.selectedManifestId];
        var manifestDetailsBox = selectedManifest
            ? React.createElement(ManifestDetails_1.default, {manifest: selectedManifest})
            : React.createElement("div", {className: ManifestManager_module_scss_1.default.noManifestSelectedContainer}, "No manifest selected");
        return React.createElement("div", {className: ManifestManager_module_scss_1.default.container}, React.createElement("div", {className: ManifestManager_module_scss_1.default.manifestListContainer}, React.createElement(ManifestList_1.default, {manifests: this.state.manifests, selectedManifestId: this.state.selectedManifestId, manifestSelectionChangedHandler: function (selectedManifestId) {
            if (selectedManifestId === _this.state.selectedManifestId) {
                selectedManifestId = undefined;
            }
            ManifestManagerActionCreator_1.default.selectManifest(selectedManifestId);
        }})), React.createElement("div", {className: ManifestManager_module_scss_1.default.manifestDetailsContainer}, manifestDetailsBox));
    };
    return ManifestManager;
}(React.Component));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ManifestManager;
