// office-ui-fabric-react/lib-amd/demo/components/App/AppState.js
define(["require", "exports", '../../pages/BreadcrumbPage/BreadcrumbPage', '../../pages/ButtonPage/ButtonPage', '../../pages/CalloutPage/CalloutPage', '../../pages/ColorPickerPage/ColorPickerPage', '../../pages/DocumentCardPage/DocumentCardPage', '../../pages/CheckboxPage/CheckboxPage', '../../pages/ChoiceGroupPage/ChoiceGroupPage', '../../pages/CommandBarPage/CommandBarPage', '../../pages/ContextualMenuPage/ContextualMenuPage', '../../pages/DatePickerPage/DatePickerPage', '../../pages/DetailsListPage/DetailsListPage', '../../pages/DialogPage/DialogPage', '../../pages/DropdownPage/DropdownPage', '../../pages/Facepile/FacepilePage', '../../pages/FocusZonePage/FocusZonePage', '../../pages/FocusTrapZonePage/FocusTrapZonePage', '../../pages/GroupedListPage/GroupedListPage', '../../pages/ImagePage/ImagePage', '../../pages/LabelPage/LabelPage', '../../pages/LayerPage/LayerPage', '../../pages/LinkPage/LinkPage', '../../pages/ListPage/ListPage', '../../pages/MarqueeSelectionPage/MarqueeSelectionPage', '../../pages/MessageBarPage/MessageBarPage', '../../pages/NavPage/NavPage', '../../pages/OrgChartPage/OrgChartPage', '../../pages/OverlayPage/OverlayPage', '../../pages/PanelPage/PanelPage', '../../pages/PeoplePickerPage/PeoplePickerPage', '../../pages/PersonaCardPage/PersonaCardPage', '../../pages/PersonaPage/PersonaPage', '../../pages/PivotPage/PivotPage', '../../pages/ProgressIndicatorPage/ProgressIndicatorPage', '../../pages/SearchBoxPage/SearchBoxPage', '../../pages/SelectionPage/SelectionPage', '../../pages/SliderPage/SliderPage', '../../pages/SpinnerPage/SpinnerPage', '../../pages/TextFieldPage/TextFieldPage', '../../pages/TogglePage/TogglePage', '../../pages/ThemePage/ThemePage'], function (require, exports, BreadcrumbPage_1, ButtonPage_1, CalloutPage_1, ColorPickerPage_1, DocumentCardPage_1, CheckboxPage_1, ChoiceGroupPage_1, CommandBarPage_1, ContextualMenuPage_1, DatePickerPage_1, DetailsListPage_1, DialogPage_1, DropdownPage_1, FacepilePage_1, FocusZonePage_1, FocusTrapZonePage_1, GroupedListPage_1, ImagePage_1, LabelPage_1, LayerPage_1, LinkPage_1, ListPage_1, MarqueeSelectionPage_1, MessageBarPage_1, NavPage_1, OrgChartPage_1, OverlayPage_1, PanelPage_1, PeoplePickerPage_1, PersonaCardPage_1, PersonaPage_1, PivotPage_1, ProgressIndicatorPage_1, SearchBoxPage_1, SelectionPage_1, SliderPage_1, SpinnerPage_1, TextFieldPage_1, TogglePage_1, ThemePage_1) {
    "use strict";
    (function (ExampleStatus) {
        ExampleStatus[ExampleStatus["placeholder"] = 0] = "placeholder";
        ExampleStatus[ExampleStatus["started"] = 1] = "started";
        ExampleStatus[ExampleStatus["beta"] = 2] = "beta";
        ExampleStatus[ExampleStatus["release"] = 3] = "release";
    })(exports.ExampleStatus || (exports.ExampleStatus = {}));
    var ExampleStatus = exports.ExampleStatus;
    exports.AppState = {
        appTitle: 'Fabric - React',
        examplePages: [
            {
                links: [
                    {
                        component: BreadcrumbPage_1.BreadcrumbPage,
                        name: 'Breadcrumb',
                        status: ExampleStatus.beta,
                        url: '#/examples/breadcrumb'
                    },
                    {
                        component: ButtonPage_1.ButtonPage,
                        name: 'Button',
                        status: ExampleStatus.beta,
                        url: '#/examples/button'
                    },
                    {
                        component: CalloutPage_1.CalloutPage,
                        name: 'Callout',
                        status: ExampleStatus.beta,
                        url: '#/examples/callout'
                    },
                    {
                        component: CheckboxPage_1.CheckboxPage,
                        name: 'Checkbox',
                        status: ExampleStatus.beta,
                        url: '#/examples/checkbox'
                    },
                    {
                        component: ChoiceGroupPage_1.ChoiceGroupPage,
                        name: 'ChoiceGroup',
                        status: ExampleStatus.beta,
                        url: '#/examples/choicegroup'
                    },
                    {
                        component: CommandBarPage_1.CommandBarPage,
                        name: 'CommandBar',
                        status: ExampleStatus.beta,
                        url: '#/examples/commandbar'
                    },
                    {
                        component: ContextualMenuPage_1.ContextualMenuPage,
                        name: 'ContextualMenu',
                        status: ExampleStatus.beta,
                        url: '#/examples/contextmenu'
                    },
                    {
                        component: DatePickerPage_1.DatePickerPage,
                        name: 'DatePicker',
                        status: ExampleStatus.beta,
                        url: '#/examples/datepicker'
                    },
                    {
                        component: DetailsListPage_1.DetailsListPage,
                        name: 'DetailsList',
                        status: ExampleStatus.beta,
                        url: '#/examples/detailslist'
                    },
                    {
                        component: DialogPage_1.DialogPage,
                        name: 'Dialog',
                        status: ExampleStatus.beta,
                        url: '#/examples/dialog'
                    },
                    {
                        component: DocumentCardPage_1.DocumentCardPage,
                        name: 'DocumentCard',
                        status: ExampleStatus.beta,
                        url: '#/examples/documentcard'
                    },
                    {
                        component: DropdownPage_1.DropdownPage,
                        name: 'Dropdown',
                        status: ExampleStatus.beta,
                        url: '#/examples/dropdown'
                    },
                    {
                        component: FacepilePage_1.FacepilePage,
                        name: 'Facepile',
                        status: ExampleStatus.started,
                        url: '#/examples/facepile'
                    },
                    {
                        component: LabelPage_1.LabelPage,
                        name: 'Label',
                        status: ExampleStatus.beta,
                        url: '#/examples/label'
                    },
                    {
                        component: LinkPage_1.LinkPage,
                        name: 'Link',
                        status: ExampleStatus.beta,
                        url: '#/examples/link'
                    },
                    {
                        component: ListPage_1.ListPage,
                        name: 'List',
                        status: ExampleStatus.beta,
                        url: '#/examples/list'
                    },
                    {
                        component: MessageBarPage_1.MessageBarPage,
                        name: 'MessageBar',
                        status: ExampleStatus.placeholder,
                        url: '#/examples/messagebar'
                    },
                    {
                        component: OrgChartPage_1.OrgChartPage,
                        name: 'OrgChart',
                        status: ExampleStatus.placeholder,
                        url: '#/examples/orgchart'
                    },
                    {
                        component: OverlayPage_1.OverlayPage,
                        name: 'Overlay',
                        status: ExampleStatus.beta,
                        url: '#/examples/overlay'
                    },
                    {
                        component: PanelPage_1.PanelPage,
                        name: 'Panel',
                        status: ExampleStatus.beta,
                        url: '#/examples/panel'
                    },
                    {
                        component: PeoplePickerPage_1.PeoplePickerPage,
                        name: 'PeoplePicker',
                        status: ExampleStatus.started,
                        url: '#/examples/peoplepicker'
                    },
                    {
                        component: PersonaPage_1.PersonaPage,
                        name: 'Persona',
                        status: ExampleStatus.beta,
                        url: '#/examples/persona'
                    },
                    {
                        component: PersonaCardPage_1.PersonaCardPage,
                        name: 'PersonaCard',
                        status: ExampleStatus.placeholder,
                        url: '#/examples/personacard'
                    },
                    {
                        component: PivotPage_1.PivotPage,
                        name: 'Pivot',
                        status: ExampleStatus.started,
                        url: '#/examples/pivot'
                    },
                    {
                        component: ProgressIndicatorPage_1.ProgressIndicatorPage,
                        name: 'ProgressIndicator',
                        status: ExampleStatus.beta,
                        url: '#/examples/progressindicator'
                    },
                    {
                        component: SearchBoxPage_1.SearchBoxPage,
                        name: 'SearchBox',
                        status: ExampleStatus.started,
                        url: '#/examples/searchbox'
                    },
                    {
                        component: SpinnerPage_1.SpinnerPage,
                        name: 'Spinner',
                        status: ExampleStatus.beta,
                        url: '#/examples/spinner'
                    },
                    {
                        component: TextFieldPage_1.TextFieldPage,
                        name: 'TextField',
                        status: ExampleStatus.beta,
                        url: '#/examples/textfield'
                    },
                    {
                        component: TogglePage_1.TogglePage,
                        name: 'Toggle',
                        status: ExampleStatus.beta,
                        url: '#/examples/toggle'
                    }
                ],
                name: 'Basic components'
            },
            {
                links: [
                    {
                        component: ColorPickerPage_1.ColorPickerPage,
                        name: 'ColorPicker',
                        status: ExampleStatus.started,
                        url: '#/examples/colorpicker'
                    },
                    {
                        component: GroupedListPage_1.GroupedListPage,
                        name: 'GroupedList',
                        status: ExampleStatus.started,
                        url: '#examples/groupedlist'
                    },
                    {
                        component: ImagePage_1.ImagePage,
                        name: 'Image',
                        status: ExampleStatus.beta,
                        url: '#/examples/image'
                    },
                    {
                        component: LayerPage_1.LayerPage,
                        name: 'Layer',
                        status: ExampleStatus.beta,
                        url: '#/examples/layer'
                    },
                    {
                        component: NavPage_1.NavPage,
                        name: 'Nav',
                        status: ExampleStatus.started,
                        url: '#/examples/nav'
                    },
                    {
                        component: SliderPage_1.SliderPage,
                        name: 'Slider',
                        status: ExampleStatus.beta,
                        url: '#/examples/slider'
                    }
                ],
                name: 'Extended components'
            },
            {
                links: [
                    {
                        component: FocusTrapZonePage_1.FocusTrapZonePage,
                        name: 'FocusTrapZone',
                        status: ExampleStatus.beta,
                        url: '#examples/focustrapzone'
                    },
                    {
                        component: FocusZonePage_1.FocusZonePage,
                        name: 'FocusZone',
                        status: ExampleStatus.beta,
                        url: '#examples/focuszone'
                    },
                    {
                        component: MarqueeSelectionPage_1.MarqueeSelectionPage,
                        name: 'MarqueeSelection',
                        status: ExampleStatus.beta,
                        url: '#examples/marqueeselection'
                    },
                    {
                        component: SelectionPage_1.SelectionPage,
                        name: 'Selection',
                        status: ExampleStatus.beta,
                        url: '#examples/selection'
                    },
                    {
                        component: ThemePage_1.ThemePage,
                        name: 'Themes',
                        url: '#examples/themes'
                    }
                ],
                name: 'Utilities'
            }
        ],
        headerLinks: [
            {
                name: 'Getting started',
                url: '#/'
            },
            {
                name: 'Fabric',
                url: 'http://dev.office.com/fabric'
            },
            {
                name: 'Github',
                url: 'http://www.github.com/officedev'
            }
        ]
    };
});
