// @microsoft/sp-tslint-rules/lib/tests/TestHelper.js
/**
 * This file is copy-pasted from https://github.com/Microsoft/tslint-microsoft-contrib/blob/master/tests/TestHelper.ts
 * Should remove this and import from tslint-microsoft-contrib package once it exports this file.
 * Tracking github issue link: https://github.com/Microsoft/tslint-microsoft-contrib/issues/184
 */
"use strict";
// tslint:disable
var Lint = require('tslint/lib/lint');
var fs = require('fs');
var chai = require('chai');
/**
 * Test Utilities.
 */
var TestHelper;
(function (TestHelper) {
    function assertNoViolation(ruleName, inputFileOrScript) {
        runRuleAndEnforceAssertions(ruleName, null, inputFileOrScript, []);
    }
    TestHelper.assertNoViolation = assertNoViolation;
    function assertNoViolationWithOptions(ruleName, options, inputFileOrScript) {
        runRuleAndEnforceAssertions(ruleName, options, inputFileOrScript, []);
    }
    TestHelper.assertNoViolationWithOptions = assertNoViolationWithOptions;
    function assertViolationsWithOptions(ruleName, options, inputFileOrScript, expectedFailures) {
        runRuleAndEnforceAssertions(ruleName, options, inputFileOrScript, expectedFailures);
    }
    TestHelper.assertViolationsWithOptions = assertViolationsWithOptions;
    function assertViolations(ruleName, inputFileOrScript, expectedFailures) {
        runRuleAndEnforceAssertions(ruleName, null, inputFileOrScript, expectedFailures);
    }
    TestHelper.assertViolations = assertViolations;
    function runRule(ruleName, userOptions, inputFileOrScript) {
        var configuration = {
            rules: {}
        };
        if (userOptions != null && userOptions.length > 0) {
            //options like `[4, 'something', false]` were passed, so prepend `true` to make the array like `[true, 4, 'something', false]`
            configuration.rules[ruleName] = [true].concat(userOptions);
        }
        else {
            configuration.rules[ruleName] = true;
        }
        var options = {
            formatter: 'json',
            configuration: configuration,
            rulesDirectory: 'lib/rules/',
            formattersDirectory: 'lib/formatters/'
        };
        var result;
        if (inputFileOrScript.match(/.*\.ts(x)?$/)) {
            var contents = fs.readFileSync(inputFileOrScript, 'utf8');
            var linter = new Lint.Linter(inputFileOrScript, contents, options);
            result = linter.lint();
        }
        else {
            var linter = new Lint.Linter('file.ts', inputFileOrScript, options);
            result = linter.lint();
        }
        return result;
    }
    TestHelper.runRule = runRule;
    function runRuleAndEnforceAssertions(ruleName, userOptions, inputFileOrScript, expectedFailures) {
        var lintResult = runRule(ruleName, userOptions, inputFileOrScript);
        var actualFailures = JSON.parse(lintResult.output);
        // All the information we need is line and character of start position. For JSON comparison
        // to work, we will delete the information that we are not interested in from both actual and
        // expected failures.
        actualFailures.forEach(function (actual) {
            delete actual.startPosition.position;
            delete actual.endPosition;
            // Editors start counting lines and characters from 1, but tslint does it from 0.
            // To make thing easier to debug, align to editor values.
            actual.startPosition.line = actual.startPosition.line + 1;
            actual.startPosition.character = actual.startPosition.character + 1;
        });
        expectedFailures.forEach(function (expected) {
            delete expected.startPosition.position;
            delete expected.endPosition;
        });
        var errorMessage = "Wrong # of failures: \n" + JSON.stringify(actualFailures, null, 2);
        chai.assert.equal(expectedFailures.length, actualFailures.length, errorMessage);
        expectedFailures.forEach(function (expected, index) {
            var actual = actualFailures[index];
            chai.assert.deepEqual(actual, expected);
        });
    }
})(TestHelper = exports.TestHelper || (exports.TestHelper = {}));
