// @microsoft/sp-client-preview/lib/local/common/dataAccess/DataLayerError.js
"use strict";
/** Represents a datalayer error */
var DataLayerError = (function () {
    function DataLayerError(errorType, correlationId, statusCode) {
        this.errorType = errorType;
        this.correlationId = correlationId;
        this.statusCode = statusCode;
    }
    DataLayerError.prototype.toString = function () {
        return "ErrorType: " + this.errorType + ", StatusCode: " + this.statusCode + ", CorrelationId: " + this.correlationId;
    };
    return DataLayerError;
}());
exports.DataLayerError = DataLayerError;
/** Represents the type of datalayer error */
(function (DataLayerErrorType) {
    /** XMLHttpRequest.onError transport level failure */
    DataLayerErrorType[DataLayerErrorType["TransportError"] = 0] = "TransportError";
    /** HTTP status code not in the 200 -> 300 range */
    DataLayerErrorType[DataLayerErrorType["ResponseError"] = 1] = "ResponseError";
    /** Error parsing the response */
    DataLayerErrorType[DataLayerErrorType["ParseError"] = 2] = "ParseError";
    /** Timeout */
    DataLayerErrorType[DataLayerErrorType["TimeOutError"] = 3] = "TimeOutError";
    /** Unknown error. This should never happen. Means unhandled error */
    DataLayerErrorType[DataLayerErrorType["UnknownError"] = 4] = "UnknownError";
})(exports.DataLayerErrorType || (exports.DataLayerErrorType = {}));
var DataLayerErrorType = exports.DataLayerErrorType;
