// @microsoft/sp-module-loader/lib/DeveloperTools/Components/DeveloperModules/PerformanceDisplay/PerformanceDisplay.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
/**
 * @file PerformanceDisplay.tsx
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 */
var React = require('react');
var PerformanceDisplay_module_scss_1 = require('./PerformanceDisplay.module.scss');
var PerformanceGraph_1 = require('./PerformanceGraph/PerformanceGraph');
var PerformanceHeader_1 = require('./PerformanceHeader/PerformanceHeader');
var PerformanceDisplayStore_1 = require('../../../Stores/PerformanceDisplayStore');
var PerformanceDisplay = (function (_super) {
    __extends(PerformanceDisplay, _super);
    function PerformanceDisplay() {
        var _this = this;
        _super.call(this);
        this._setState = function () {
            _this.setState({
                perfItems: PerformanceDisplayStore_1.default.performanceData,
                eupl: PerformanceDisplayStore_1.default.eupl,
                startTime: PerformanceDisplayStore_1.default.startTime,
                errorMessage: PerformanceDisplayStore_1.default.errorMessage
            });
        };
        this.state = {
            perfItems: PerformanceDisplayStore_1.default.performanceData,
            eupl: PerformanceDisplayStore_1.default.eupl,
            startTime: PerformanceDisplayStore_1.default.startTime,
            errorMessage: PerformanceDisplayStore_1.default.errorMessage
        };
    }
    PerformanceDisplay.prototype.componentDidMount = function () {
        PerformanceDisplayStore_1.default.addListener(this._setState);
    };
    PerformanceDisplay.prototype.componentWillUnmount = function () {
        PerformanceDisplayStore_1.default.removeListener(this._setState);
    };
    PerformanceDisplay.prototype.render = function () {
        var errorMessage = [PerformanceDisplay_module_scss_1.default.errorMessage, PerformanceDisplay_module_scss_1.default.msFontColorRedDark].join(' ');
        // @todo (VSO bug 236724): localize these strings
        if (this.state.errorMessage) {
            return (React.createElement("div", {className: errorMessage}, "Sorry, something went wrong.", React.createElement("br", null), "Error Message: ", this.state.errorMessage));
        }
        else {
            return (React.createElement("div", {className: PerformanceDisplay_module_scss_1.default.performanceDisplayContainer}, React.createElement(PerformanceHeader_1.default, {perfItems: this.state.perfItems, overallDuration: this.state.eupl}), React.createElement("div", {className: PerformanceDisplay_module_scss_1.default.graphContainer}, React.createElement(PerformanceGraph_1.default, {perfItems: this.state.perfItems, startTime: this.state.startTime, overallDuration: this.state.eupl}))));
        }
    };
    return PerformanceDisplay;
}(React.Component));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PerformanceDisplay;
